modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/restapi/utils/RESTAPITestUtil.java

```diff
@@ -0,0 +1,377 @@
+     *
+     * @param configFilePath location of the data file
+     * @param gatewayURL     gateway URL
+     * @param keyMangerURL   key manager URL
+     * @return true if all the scenarios in the current test is successful, otherwise false
+     */
+    public boolean testRestAPI(String configFilePath, String gatewayURL, String keyMangerURL) {
+
+        //this is the boolean which indicates the success or failure of the current test
+        boolean isTestSuccess = false;
+
+        try {
+
+            String configData = getConfigurations(configFilePath);
+            Map<String, String> dataMap = registerOAuthApplication(keyMangerURL);
+            String accessToken;
+            JSONObject testScenario = new JSONObject(configData);
+            JSONArray array = testScenario.getJSONArray(RESTAPITestConstants.JSON_ROOT_ELEMENT);
+            int arrayLength = array.length();
+            HashMap<String, String> preservedAttributes = new HashMap<String, String>();
+
+            //go through each section of the JSON string and fetch data
+            for (int i = 0; i < arrayLength; i++) {
+
+                JSONObject configObject = (JSONObject) array.get(i);
+                JSONObject initializationConfigurationObject =
+                        configObject.getJSONObject(RESTAPITestConstants.INITIALIZATION_SECTION);
+                String scope = initializationConfigurationObject.get(RESTAPITestConstants.SCOPE_ELEMENT).toString();
+
+                if (dataMap == null || dataMap.isEmpty()) {
+                    //if the data map is null or empty, that means it has failed to register OAuth application
+                    log.error("Failed to register OAuth application to test the REST API.");
+                    isTestSuccess = false;
+                    //stop the rest of the iterations of the loop if it has failed to register OAuth application
+                    break;
+                }
+
+                accessToken = generateOAuthAccessToken(scope, dataMap, gatewayURL);
+
+                if (accessToken == null) {
+                    //if the access token is null, that means it has failed to generate OAuth access token
+                    log.error("Failed to generate OAuth access token to test the REST API.");
+                    isTestSuccess = false;
+                    //stop the rest of the iterations of the loop if the test has failed due to accessToken error
+                    break;
+                }
+
+                JSONObject dataConfigurationObject = configObject.getJSONObject(RESTAPITestConstants.DATA_SECTION);
+                String method = dataConfigurationObject.get(RESTAPITestConstants.METHOD_ELEMENT).toString();
+                String resourceUrl = keyMangerURL +
+                        dataConfigurationObject.get(RESTAPITestConstants.URL_ELEMENT).toString();
+                Pattern parameterPattern = Pattern.compile("\\{(.*?)\\}");
```

Comment:
Pattern compilation should be done off the critical path of a request. Initialize and compile the pattern when the class is loaded and re-use. This should be possible since Pattern instances are thread-safe.